crates/red_knot_python_semantic/resources/mdtest/annotations/never.md

```diff
@@ -0,0 +1,57 @@
+# NoReturn & Never
+
+## Annotation
+
+`NoReturn` to annotate functions that never return normally. `Never` represents the bottom type, a
+type that represents the empty set of Python objects. These two annotations can be used
+interchangeably.
+
+```py
+from typing import NoReturn, Never, Any
+
+# error: [invalid-type-parameter] "Type `typing.Never` expected no type parameter"
+x: Never[int]
+
+def stop() -> NoReturn:
+    raise RuntimeError("no way")
+
+# revealed: NoReturn
+reveal_type(stop())
```

Comment:
I think this should be a separate test that ends here. If you were to actually run this code, the program would never return from the `stop()` call.

Everything after this line can be considered dead code from the perspective of static analysis. Which also means we might not (should not) emit any diagnostics, including `reveal_type`-information. I guess it's even questionable as to whether or not the `reveal_type` in this line should emit a diagnostic. You would not see any result at runtime. Existing type checkers still reveal the type for the `stop()` call, but nothing beyond.

Reply:
My intention was to test the types of these expressions. You're right this test will break when the behavior for Never is added.
I will move this call and other `reveal_types` that can stop the type checking to a function.

> I think this should be a separate test that ends here.

Are you suggesting to implement this behavior in this PR? Otherwise I can add a todo comment here and work on it in a follow up PR.

Reply:
> Are you suggesting to implement this behavior in this PR?

Ah, no — sorry. I was merely suggesting that we split this large test into multiple smaller tests. And take care not to have any test assertions (`# revealed: ` / `# error: `) in lines following a call to a function that returns `Never`/`NoReturn`. To make sure this test will work in the future once we add unreachable-code analysis.

Writing smaller self-consistent tests would also be beneficial if we ever need to debug an assertion failure in this file.